partiql-plan/src/main/java/org/partiql/plan/OperatorTransform.java

```diff
@@ -0,0 +1,695 @@
+package org.partiql.plan;
+
+import org.jetbrains.annotations.NotNull;
+import org.partiql.plan.rel.*;
+import org.partiql.plan.rel.RelAggregate.Measure;
+import org.partiql.plan.rex.*;
+import org.partiql.plan.rex.RexCase.Branch;
+import org.partiql.plan.rex.RexStruct.Field;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Operator transform is an abstract base visitor which recursively rewrites an operator tree.
+ */
+public abstract class OperatorTransform<C> implements OperatorVisitor<Operator, C> {
+
+    /**
+     * Operator factory to use for rewrites.
+     */
+    private final Operators operators;
+
+    /**
+     * Base operator transform with standard operators factory.
+     */
+    public OperatorTransform() {
+        this.operators = Operators.STANDARD;
+    }
+
+    /**
+     * Base operator transform with custom operators factory.
+     */
+    public OperatorTransform(Operators operators) {
+        this.operators = operators;
+    }
+
+    @NotNull
+    @Override
+    public Operator defaultReturn(@NotNull Operator operator, C ctx) {
+        return operator;
+    }
+
+    @NotNull
+    @Override
+    public Operator defaultVisit(Operator operator, C ctx) {
+        return operator;
+    }
+
+    @NotNull
+    @Override
+    public Operator visitAggregate(@NotNull RelAggregate rel, C ctx) {
+        // rewrite input
+        Rel input = rel.getInput();
+        Rel input_new = visit(input, ctx, Rel.class);
+        // rewrite measures
+        List<Measure> measures = rel.getMeasures();
+        List<Measure> measures_new = visitAll(measures, ctx, this::visitAggregateMeasure);
+        // rewrite groups
+        List<Rex> groups = rel.getGroups();
+        List<Rex> groups_new = visitAll(groups, ctx, this::visitAggregateGroup);
+        // rewrite aggregate
+        if (input != input_new || measures != measures_new || groups != groups_new) {
```

Comment:
One thing to highlight is that Java's `!=` will use referential equality, which wouldn't take into consideration the methods that now make plan nodes mutable such as `setTag()` and `setType()`.

Reply:
We want referential equality for a transform. You'll see all the pig rewriters and sprout rewriters use kotlin referential equality. 

You don't need a rewriter for mutable things because mutable.